spylon/spark/progress.py

```diff
@@ -0,0 +1,146 @@
+    elif minutes > 0:
+        return '{sign}{minutes:02d}m:{seconds:02d}s'.format(**d)
+    else:
+        return '{sign}{seconds:02d}s'.format(**d)
+
+
+def spark_progress_thread_worker(status, timedelta_formatter=pretty_time_delta, bar_width=20, sleep_time=0.5):
+    """
+
+    Parameters
+    ----------
+    status : :class:`pyspark.status.StatusTracker`
+    timedelta_formatter : function
+        callable that converts a timedelta to a string.
+    bar_width : int
+        width of the progressbar to print out.
+    sleep_time : float
+        Frequency with which to poll spark for new progress information.
+
+    """
+    last_status = ''
+    start_times = defaultdict(datetime.datetime.now)
+    max_stage_id = 0
+
+    while True:
+        stage_ids = status.getActiveStageIds()
+        progressbar_list = []
+        # Only show first 3
+        for sid in stage_ids[:3]:
+            info = status.getStageInfo(sid)
+            if info:
+                td = datetime.datetime.now() - start_times[sid]
+                dur = timedelta_formatter(td)
+
+                percent = (info.numCompletedTasks * bar_width) // info.numTasks
+                bar = [' '] * bar_width
+                for i in range(bar_width):
+                    char = ' '
+                    if i < percent:
+                        char = '='
+                    if i == percent:
+                        char = '>'
+                    bar[i] = char
+                bar = ''.join(bar)
+
+                s = "[Stage {sid}:{bar} " \
+                    "({info.numCompletedTasks} + {info.numActiveTasks} / {info.numTasks} Dur: {dur}]"\
+                    .format(sid=sid, info=info, dur=dur, bar=bar)
+                progressbar_list.append(s)
+
+        if len(stage_ids):
+            current_max_stage = max(stage_ids[:3])
+
+            # Ensure that when we get a new maximum stage id we print a \n to make the progress bar go on to the next
+            # line.
+            if current_max_stage > max_stage_id:
+                if last_status != '':
+                    sys.stderr.write("\n")
+                sys.stderr.flush()
+                max_stage_id = current_max_stage
+
+        new_status = ' '.join(progressbar_list)
+        if new_status != last_status:
+            sys.stderr.write("\r" + new_status)
+            sys.stderr.flush()
+            last_status = new_status
+        time.sleep(sleep_time)
+
+
+_progressbar_thread_started = False
+
+
+def start_spark_progress_bar_thread(sc, **kwargs):
+    """Starts a background thread that polls the spark context for progress on stages that are running.
+
+    This makes it simple to have progress bars for spark jobs inside Jupyter notebooks.
+
+    Parameters
+    ----------
+    sc : :class:`SparkContext`
+    """
+    global _progressbar_thread_started
+
+    if _progressbar_thread_started:
+        raise Exception("Spark progress thread already running")
```

Comment:
maybe you want a RuntimeException here? Exception seems a little vague (and the code that catches this Exception has a bad code smell)